File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return validate(root, -1L + Integer.MIN_VALUE, 1L + Integer.MAX_VALUE);
    }

    private boolean validate(TreeNode root, long minBoundary, long maxBoundary)
    {
        if (root == null) return true;

        if (root.val <= minBoundary || root.val >= maxBoundary) return false;

        return validate(root.left, minBoundary, root.val) && validate(root.right, root.val, maxBoundary);
    }

}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def validate(root, minBoundary, maxBoundary):
            if not root: return True

            if root.val <= minBoundary: return False
            if root.val >= maxBoundary: return False

            return validate(root.left, minBoundary, root.val) & validate(root.right, root.val, maxBoundary)


        return validate(root, -pow(2, 31) -1, pow(2, 31))
        